### DISCUSSION TITLE
Do you still need to zero-initialize stackalloc using Span<T>.Clear()?

### DISCUSSION THREAD

**Discussion by samuel-lucas6** [dotnet/runtime#74860] [Q&A] [answered]
From [this](https://vcsjones.dev/stackalloc/) blog post back in 2020:

> DON’T: Assume stack allocations are zero initialized
> Most normal uses of stackalloc result in zero-initialized data. This behavior is however not guaranteed, and can change depending if the application is built for Debug or Release, and other contents of the method. Therefore, don’t assume that any of the elements in a stackalloced Span<T> are initialized to something by default.

The stackalloc documentation also says:

> The content of the newly allocated memory is undefined. You should initialize it before the use. For example, you can use the [Span<T>.Clear](https://docs.microsoft.com/en-us/dotnet/api/system.span-1.clear) method that sets all the items to the default value of type T.

However, I haven't come across a case where it hasn't been zero-initialized, and I thought that example (Debug vs Release) was fixed. Is this still necessary in .NET 6 and the upcoming 7?

**Comment by PathogenDavid** [comment, marked as answer]
As you've read, the only guarantee you get is that there is no guarantee. You don't necessarily need to use `Span.Clear`, but you need to ensure you don't read data which wasn't previously written.

----------

In practice though, locals (including stack allocations) will always be initialized to zero unless you use [`SkipLocalsInit`](https://github.com/dotnet/csharplang/blob/c82c50785ab8ed4a43805408cb152931eb8c0c12/proposals/csharp-9.0/skip-localsinit.md), which is why you aren't ever seeing it as non-zero. (I don't think it has ever actually differed between debug and release builds.)

However if you're relying on this, you're still relying on unspecified behavior. You're also creating a ticking time bomb if anyone were to apply `SkipLocalsInit` project-wide. (Which is not uncommon in performance-minded code bases.)

If you *do* want to rely on the stackalloc being initialized to zero and you don't want to pay the cost of double-initializing it: You should explicitly clear it and apply `SkipLocalsInit` to your method.
